File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        leng = 0
        temp_head = head
        while temp_head:
            leng += 1
            temp_head = temp_head.next
        idx_remove = leng - n + 1
        cur_idx = 1
        dummy = ListNode()
        res = dummy
        while head:
            if cur_idx != idx_remove:
                res.next = head
                res = res.next
            head = head.next
            cur_idx += 1
        
        res.next = None
        return dummy.next

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* node1, ListNode* node2){
        ListNode res = ListNode(0);
        ListNode* dummy = &res;
        while (node1 != nullptr && node2 != nullptr){
            if (node1->val <= node2->val){
                dummy->next = node1;
                node1 = node1->next;
                dummy = dummy->next;
            }
            else {
                dummy->next = node2;
                node2 = node2->next;
                dummy = dummy->next;
            }
        }
        if (node1 != nullptr){
            dummy->next = node1;
        }
        if (node2 != nullptr){
            dummy->next = node2;
        }
        return res.next;
    }
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        int listsSize = static_cast<int>(lists.size());
        if (listsSize == 0){
            return {};
        }

        if (listsSize== 1){
            return lists[0];
        }
        vector<ListNode*> mergedLists;
        for (int i = 0; i < listsSize; i = i + 2){
            if (i + 1 < listsSize){
                mergedLists.push_back(mergeTwoLists(lists[i], lists[i + 1]));
            }
            else {
                mergedLists.push_back(mergeTwoLists(lists[i], nullptr));
            }
        }
        return mergeKLists(mergedLists);
    }
};
